install/0000_90_cluster-version-operator_02_servicemonitor.yaml

```diff
@@ -87,9 +87,9 @@ spec:
     - alert: ClusterOperatorDown
       annotations:
         summary: Cluster operator has not been available for 10 minutes.
-        description: The {{ "{{ $labels.name }}" }} operator may be down or disabled, and the components it manages may be unavailable or degraded.  Cluster upgrades may not complete. For more information refer to 'oc get -o yaml clusteroperator {{ "{{ $labels.name }}" }}'{{ "{{ with $console_url := \"console_url\" | query }}{{ if ne (len (label \"url\" (first $console_url ) ) ) 0}} or {{ label \"url\" (first $console_url ) }}/settings/cluster/{{ end }}{{ end }}" }}.
+        description: The {{ "{{ $labels.name }}" }} operator may be down or disabled because {{ "${{ $labels.reason }}" }}, and the components it manages may be unavailable or degraded.  Cluster upgrades may not complete. For more information refer to 'oc get -o yaml clusteroperator {{ "{{ $labels.name }}" }}'{{ "{{ with $console_url := \"console_url\" | query }}{{ if ne (len (label \"url\" (first $console_url ) ) ) 0}} or {{ label \"url\" (first $console_url ) }}/settings/cluster/{{ end }}{{ end }}" }}.
       expr: |
-        max by (namespace, name) (cluster_operator_up{job="cluster-version-operator"} == 0)
+        max by (namespace, name, reason) (cluster_operator_up{job="cluster-version-operator"} == 0)
```

Comment:
> I am not clear how much cardinality increase we are looking at by adding reason.

We already have `reason` cardinality on `cluster_operator_conditions`, and that seems ok.  Checking 4.11 Telemetry:

```
sort_desc(
  count by (name) (
    count by (name, reason) (
      (cluster_operator_conditions{condition="Available"} == 0)
      + on (_id) group_left ()
      group by (_id) (cluster_version{type="current",version=~"4[.]11[.][0-9]*"})
    )
  )
)
```

gives:

```
{name="authentication"}  16
{name="monitoring"}  10
{name="storage"}  6
{name="console"}  5
{name="image-registry"}  3
...
```

so the cardinality increase would be mostly limited to those particular operators.

`message` cardinality will be higher, potentially including things like request port numbers, API response IDs, and that sort of thing.

> I am still not clear about the motivation.

Cluster admins like SRE-P get a `ClusterOperatorDown` alert.  They wonder if it's a true-positive (actual cluster pain; needs quick manual intervention) or a false-positive (operator is over-reacting, cluster is actually ok, or will soon self-heal to become ok).  Splitting `ClusterOperatorDown` by `name` gives you some granularity (e.g. you can decide whether a particular component like `etcd` is often a true-positive or often a false-positive.  But splitting by `(name, reason)` tuples gives more granularity, so you can more precisely separate likely-to-be-a-false-positive from likely-to-be-a-true-positive.

Reply:
> Cluster admins like SRE-P get a ClusterOperatorDown alert. They wonder if it's a true-positive (actual cluster pain; needs quick manual intervention) or a false-positive (operator is over-reacting, cluster is actually ok, or will soon self-heal to become ok). Splitting ClusterOperatorDown by name gives you some granularity (e.g. you can decide whether a particular component like etcd is often a true-positive or often a false-positive. But splitting by (name, reason) tuples gives more granularity, so you can more precisely separate likely-to-be-a-false-positive from likely-to-be-a-true-positive.

I'm not entirely convinced about this. Do we want to optimize for what is basically an OpenShift bug (operator down should never be a false-positive thing to be ignored, right?). I feel that we'd just add some tech jargon to admin-facing messages, after admin-ing the CI clusters myself for some time I'm not sure if I found that too useful tbh.

That said, it does not seem the cost of adding the labels is too high, and I can imagine they can be useful when troubleshooting, diagnosis or maybe conditional risks.

Reply:
> It would (maybe) be better to not silence the alert and let CAD (Cluster Anomaly Detection) deal with the alert and send the proper SL to the customer.

An issues with automatically service-logging on these alerts is that, while sometimes they are things the customer can fix (hooray). sometimes they are things that SRE-P should fix for the customer (that's why customers pay us to manage clusters).  For example:

* Sometimes the alert is really a false-positive.  The appropriate action is to file an OCPBUGS ticket, which then gets fixed, and the fixes ship, and then clusters get updated to pick up the fix.  That can still be a long turnaround time, and because some managed offerings like OSD defer control-plane patch updates to the customer, the turnaround time can be _very_ slow.  Silences give service delivery something to de-noise in the meantime.  Customers don't want to be bothered by this loop.
* Sometimes the alert is a true-positive, but it's something SRE-P is supposed to fix without bothering the customer.  E.g. `monitoring`'s `UpdatingPrometheusOperatorFailed` is probably not something that the customer wants to be bothered about.